.github/workflows/Tests.yml

```diff
@@ -457,6 +463,11 @@
       if: ${{ runner.os }} != "Linux"
       run: |
         if [ "$OS" == "macos-latest" ] ; then ./cc-test-reporter after-build --exit-code 0 || true ; else echo "::warning file=.github/workflows/Tests.yml,line=459,endLine=459,title=SKIPPED::SKIP Code-Climate" ; fi ;
+      if: ${{ !cancelled() }}
+    - name: code-climate for ${{ matrix.python-version }}
+      if: ${{ runner.os }} == "Linux"
+      run: |
+        ./cc-test-reporter after-build --exit-code 0 || true ;
```

Comment:
LGTM, it does seem to evaluate to true even on windows-latest ... is there any official documentation that breaks down the syntax? I've found in the past some of github's own examples misleading regarding the `{{ ... }}` syntax of GHA configs.

🤔 Until I understand the syntax I'm going to hold-off on worsening the already significant churn around conditionals in the '.github/workflows/Tests.yml' config. The ``|| true`` trick works around the issue currently.

Reply:
Thanks, I've read up on the documentation you mentioned and I think I understand it better. I was thinking of it before more like variable expansion than as wrapping expressions.

I'll try out the `if: ${{ !cancelled() && runner.os == 'Linux' }}` fix in my next push , there are some other changes needed anyway, so zero churn is unavoidable 🤷🏻

Reply:
so while the conditionals seem to now work correctly, for example: https://github.com/reactive-firewall/multicast/actions/runs/11025864397/job/30621472807?pr=127

I've identified a failing test-case `test_sdist_includes_required_files` when the `make MANIFEST.in` (direct) / `make build` (indirect) step has never been performed. it seems in at-least python 3.12 (and possibly others) the `python3 setup.py sdist --formats=gztar` strategy for building the tar package in `tests.TestManifestInclusion.test_sdist_includes_required_files()` does not consider that building depends on `make MANIFEST.in`

example CI run with test failure in logs: https://github.com/reactive-firewall/multicast/actions/runs/11025864397/job/30621436589?pr=127

read `tests/test_manifest.py` and help me debug `test_sdist_includes_required_files`

From testing locally I know:
`make build` called before `make test` resolves the issue by creating the `MANIFEST.in` file which does not seem to be removed by `setup.py clean --all`.
You can read "Makefile" for details on what these targets do.

My idea to fix this is to updated the test targets `test`, `test-pytest` and `test-tox` to be dependent on `MANIFEST.in`, this should result in the MANIFEST.in file being picked-up properly without needing a full build step before starting the testing. please check my plan, and point out any flaws.

Also please generate a possible alternative solution, and compare it to my plan in detail.

Reply:
:thinking: Thanks, after considering your plan comparison, I have some feedback:
* as this test (test_sdist_includes_required_files) is already coupled to the build process, the additional slight coupling between build and test processes is acceptable. ±0
* The overhead of `make test` dependent on `make MANIFEST.in` is less overhead than calling `make MANIFEST.in` in "test_sdist_includes_required_files" and, as would be necessary for accurate testing, other tests in the class `TestManifestInclusion`. A single re-creation of "MANIFEST.in" (relying on the well known efficiency of `make` to prevent unnecessary re-creation of "MANIFEST.in") as done in my plan should be negligible. ±0
* my approach is simpler. +1
* my python tests remain pythonic by never calling non-python tools like `make` while testing. this preserves the status quo where make can spawn python processes but python processes only spawn other python processes (as the coverage command is already python, all paths to subprocess remain in python scope only). the Alternative plan introduces access to make from python testing, this complicates the tests AND security considerations. +2
* my solution actually keeps the invocation of `make MANIFEST.in` to a minimum by the same logic as with the overhead: one invocation or less per invocation of `make test` rather than per `TestManifestInclusion` test-suite or more (considering there are other test-suites for the build process) ±1

:thinking: considering the security implications alone, my plan is the better choice. * I'll implement my plan first, and unless it fails to resolve the problem, push it to test in CI, properly.
* Please generate a summary of this problem and my plan from our discussion, then please open a new GH issue, with the summary as the issue description and the issue label "bug".